### PR TITLE
Email features

### PR DESCRIPTION
**Pull request by japesinator**
This adds a bunch of patches from old emails sent regarding 161. It does not address all the feature requests/bug reports therein, but it does integrate all sent code.

### CODE REVIEW

**Comment by dguido** [comment]
Does this resolve https://github.com/trailofbits/onesixtyone/issues/1?

**Comment by japesinator** [comment]
#1 is resolved by #2

File: onesixtyone.c [hunk 1]
```diff
@@ -80,15 +83,22 @@ struct {
 
 void usage()
 {
+  int i;
```
Comment (offlinemark) [inline onesixtyone.c]: nitpick, maybe make this `size_t` instead of `int`

File: onesixtyone.c [hunk 2]
```diff
@@ -204,6 +276,10 @@ void init_options(int argc, char *argv[])
 			case 'o' :	o.log = 1;
 						strncpy(log_filename, optarg, sizeof(log_filename));
 						break;
+      case 'p' :	o.port = atoi(optarg);
```
Comment (offlinemark) [inline onesixtyone.c]: let's use `strtol` and add an error check to catch if they did something like `-p bogus`

File: onesixtyone.c [hunk 3]
```diff
@@ -121,11 +131,71 @@ void read_communities(char* filename)
...
+  struct in_addr startaddr;
+  struct in_addr endaddr;
+  int netmask;
+
+  addr = strdup(ipmask);
+  slash = strchr(addr, '/');
+  if (slash == NULL) {
+    netmask = 32;
+  } else {
+    netmask = atoi(slash + 1);
```
Comment (mossberg) [inline onesixtyone.c]: same here, we should use `strtol` instead of `atoi`

File: onesixtyone.c [hunk 4]
```diff
@@ -121,11 +131,71 @@ void read_communities(char* filename)
...
+{
+  int ret = -1;
+  char *addr = NULL;
+  char *slash;
+  struct addrinfo hints;
+  struct addrinfo *result = NULL;
+  unsigned long longtmp, i, ips;
+  struct in_addr startaddr;
+  struct in_addr endaddr;
+  int netmask;
```
Comment (offlinemark) [inline onesixtyone.c]: let's initialize this to 32, so we can remove the `if (slash == NULL)` below

File: onesixtyone.c [hunk 5]
```diff
@@ -121,11 +131,71 @@ void read_communities(char* filename)
...
 }
 
+int add_host(const char *ipmask)
+{
+  int ret = -1;
+  char *addr = NULL;
+  char *slash;
+  struct addrinfo hints;
+  struct addrinfo *result = NULL;
+  unsigned long longtmp, i, ips;
```
Comment (offlinemark) [inline onesixtyone.c]: let's declare `longtmp` as `in_addr_t`, since that's how it's used below

File: onesixtyone.c [hunk 6]
```diff
@@ -121,11 +131,71 @@ void read_communities(char* filename)
...
+  hints.ai_canonname = NULL;
+  hints.ai_addr = NULL;
+  hints.ai_next = NULL;
+
+  if (getaddrinfo(addr, NULL, &hints, &result) != 0) {
+    goto OUT;
+  }
+
+  longtmp = ntohl(((struct sockaddr_in *)result->ai_addr)->sin_addr.s_addr);
+  startaddr.s_addr = longtmp & (unsigned long) (0 - (1<<(32 - netmask)));
```
Comment (offlinemark) [inline onesixtyone.c]: let's change longtmp to `in_addr_t` and remove the `(unsigned long)` casts on these 2 lines

Comment (offlinemark) [reply onesixtyone.c]: and rename longtmp

**Comment by japesinator** [comment]
I've made all the suggested modifications to the codebase

**Review by offlinemark — APPROVED** [review]
looks fine!

**Pull request merged**